File: atcoder.jp/C++ (GCC 9.2.1)/abc262/abc262_c.cpp
#include <iostream>
#include <vector>
using namespace std;

int main()
{
    int N;
    cin >> N;
    vector<int> A(N);
    for (int &a : A) cin >> a;

    int cnt = 0;
    int ans = 0;
    for (int j = 1; j <= N; j++) {
        if (j == A.at(j-1)) {
            ans += cnt;
            cnt++;
        } else {
            int i = A.at(j-1);
            if (min(A.at(i-1), A.at(j-1)) == i && max(A.at(i-1), A.at(j-1)) == j) {
                ans++;
            }
        }
    }

    cout << ans << endl;

    return 0;
}

File: atcoder.jp/Rust (1.42.0)/abc262/abc262_c.rs
use proconio::{fastout, input};
use std::cmp::{min, max};

#[fastout]
fn main() {
    input! {
        n: usize,
        a: [usize; n],
    }

    let mut cnt = 0i64;
    let mut ans = 0i64;

    for j in 0..n {
        if a[j] == j+1 {
            ans += cnt;
            cnt += 1;
        } else {
            let i = a[j] - 1;
            if min(a[i], a[j]) == i+1 && max(a[i], a[j]) == j+1 {
                ans += 1;
            }
        }
    }

    println!("{}", ans);
}   
